<a href="https://colab.research.google.com/github/Uluc/data_science_and_analytics/blob/main/model_accuracy_tests.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***

In [ ]:
%pylab inline
import warnings
warnings.filterwarnings('ignore')

Populating the interactive namespace from numpy and matplotlib


In [ ]:
from sklearn.datasets import load_files
from sklearn.model_selection import train_test_split

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, f1_score 

### The movie review dataset

In [ ]:
import cloudpickle as cp
from urllib.request import urlopen
dataset = cp.load(urlopen("https://drive.google.com/uc?export=download&id=1tqjekAEy_SM_sJUvjIBbajp6I-_WmHgj"))

In [ ]:
print(len(dataset.target))
print(dataset.data[220])

2000
at the outset of swordfish , john travolta's gabriel shear is pontificating about the status of american cinema today . 
basically , he says , it boils down to a lack of imagination among the majority of writers . 
how ironic , as travolta seems to be describing his latest venture . 
swordfish is loud , violent and amoral . 
it has the audacity to justify murder and mayhem in the name of sustaining our way of life . 
and how does travolta's gabriel plan to do this ? 
by robbing billions from his own government and using the funds to out-terrorize terrorists . 
swordfish is a very cynical movie . 
it relies on an audience's perception of our leaders as ineffectual and duplicitous and on terrorists as non-human , faceless entities not worthy of compassion or consideration . 
the movie's plot is preposterous with enough illogical leaps that if the film ever slowed down , you'd actually see how ridiculous it all is . 
this is a live-action road runner cartoon , moving so quickly that 

In [ ]:
docs_train, docs_test, y_train, y_test = train_test_split(dataset.data, dataset.target, test_size=0.25, random_state=123)

Task 1 
---
Use TfidfVectorizer to fit_transform the training data (docs_train) and then transform the test data (docs_test)

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(min_df=3, max_df=0.95)
train_data = tfidf.fit_transform(docs_train)
test_data = tfidf.transform(docs_test)

Task 2 
---
Build a SVC model to predict whether a movie review is positive or negative. Test the model accuracy on the test data. Try different values (0.01, 0.1, 1, 10) for the hyper parameter 'C' and print out the model accuracy for each of the parameter value.

In [ ]:
from sklearn.svm import SVC
from sklearn.pipeline import Pipeline

values = [0.01, 0.1, 1, 10]
for c in values:

    model = SVC(kernel='linear', C=c)
    m = Pipeline([('vect', tfidf), ('svc', model)])
    m.fit(docs_train, y_train)
    y_pred = m.predict(docs_test)
    print("C: ", c)
    print("Accuracy Score: ", accuracy_score(y_test, y_pred))

C:  0.01
Accuracy Score:  0.492
C:  0.1
Accuracy Score:  0.59
C:  1
Accuracy Score:  0.844
C:  10
Accuracy Score:  0.842


Task 3
---
Naive Bayes is a prediction model based on applying Bayes’ theorem with the “naive” assumption of conditional independence between every pair of features given the value of the class variable.

Build a Naive Bayes model to predict whether a movie review is positive or negative. Test the model accuracy on the test data.

In [ ]:
from sklearn.naive_bayes import MultinomialNB
bayes = MultinomialNB(alpha=0.59)
bayes.fit(train_data, y_train)
y_pred = bayes.predict(test_data)
print("Accuracy Score: ",accuracy_score(y_test, y_pred))

Accuracy Score:  0.81


Task 4
---
A random forest is a ensemble (collective) model that fits a number of decision tree classifiers on various sub-samples of the dataset and uses averaging to improve the predictive accuracy and control over-fitting.

Build a random forest model to predict whether a movie review is positive or negative. Test the model accuracy on the test data. Try different values (20, 100, 500) for the hyper parameter 'n_estimators', i.e., the number of decision trees in the ensemble, and print out the model accuracy for each of the parameter value.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

values = [20, 100, 500]

for i in values :
  random_forest = RandomForestClassifier(max_depth=2, random_state=0, n_estimators=i)
  random_forest.fit(train_data, y_train)
  y_pred = random_forest.predict(test_data)
  print("Accuracy Score: ",accuracy_score(y_test, y_pred))
  

Accuracy Score:  0.65
Accuracy Score:  0.73
Accuracy Score:  0.76


### From the above tasks, you can observe that different models and different choice of hyper-parameter values can lead to quite different prediction performance. What is the model (and hyper-parameter) among the above that gives the best prediction? What is the worst?

SVC Model on C 1 Gives the best prediction at 0.844


SVC Model on C 0.01 Gives the worst prediction at 0.492